azure-proto-core/ArmClientOptions.cs

```diff
@@ -0,0 +1,31 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+
+namespace azure_proto_core
+{
+    public class ArmClientOptions
+    {
+        private Dictionary<Type, object> _overrides = new Dictionary<Type, object>();
+        private static readonly object _overridesLock = new object();
+
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public object GetOverrideObject<T>(Func<object> ctor)
+        {
+            object overrideObject;
+            Type type = typeof(T);
+            if (!_overrides.TryGetValue(type, out overrideObject))
```

Comment:
Is the purpose of restricting this to only being made once so that we can't have two api versions or something in effect at once? Since, if the type is already made, we never want to make it again, we don't call ctor()

Reply:
correct, this is basically giving us a singleton per instance of ArmClientOptions, but still allows each instance of ArmClientOptions to have different versions